source/agora/common/TransactionData.d

```diff
@@ -0,0 +1,284 @@
+
+    public this (scope const(char)[] hexstr) pure nothrow @nogc @safe
+    {
+        if (hexstr[0] == '0' && (hexstr[1] == 'x' || hexstr[1] == 'X'))
+            hexstr = hexstr[2 .. $];
+
+        if (hexstr.length / 2 > MAX_SIZE)
+            this.is_oversize = true;
+        else
+            this.is_oversize = false;
+
+        auto range = hexstr.byChar.map!(std.ascii.toLower!(char)).map!(parseHex);
+        size_t offset;
+        foreach (chunk; range.chunks(2))
+        {
+            (() @trusted {
+                this.data[offset++] = cast(ubyte)((chunk[0] << 4) + chunk[1]);
+            })();
+            if (offset == MAX_SIZE)
+                break;
+        }
+        this.length = offset;
+    }
+
+    /// Print the transaction data
+    public void toString (scope void delegate(const(char)[]) @safe sink) const @safe
+    {
+        static immutable hex_digits = `0123456789abcdef`;
+
+        sink("0x");
+        char[2] hex;
+        this.data[0 .. this.length].each!(
+            (elem)
+            {
+                hex[0] = hex_digits[elem >> 4];
+                hex[1] = hex_digits[(elem & 0b0000_1111)];
+                sink(hex);
+            });
+    }
+
+    /// Support for Vibe.d serialization to JSON
+    public string toString () const @safe
+    {
+        size_t idx;
+        char[] buffer;
+        buffer.length = this.length * 2 + 2;
+        scope sink = (const(char)[] v) {
+            buffer[idx .. idx + v.length] = v;
+            idx += v.length;
+        };
+        this.toString(sink);
+        return buffer.idup;
+    }
+
+    /// Parse hex string
+    public static ubyte parseHex (char c) pure nothrow @nogc @safe
+    {
+        if (c >= '0' && c <= '9')
+            return cast(ubyte)(c - '0');
+        if (c >= 'a' && c <= 'f')
+            return cast(ubyte)(10 + c - 'a');
+        assert(0, "Unexpected char in string passed to TransactionData");
+    }
+
+    /// Support for Vibe.d deserialization
+    public static TransactionData fromString (scope const(char)[] str) pure nothrow @nogc @safe
+    {
+        return TransactionData(str);
+    }
+
+    /***************************************************************************
+
+        Implements hashing support
+
+        Params:
+            dg = Hashing function accumulator
+
+    ***************************************************************************/
+
+    void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        if (this.length > 0)
+            hashPart(this.data[0 .. this.length], dg);
+    }
+
+    /***************************************************************************
+
+        TransactionData Serialization
+
+        Params:
+            dg = serialize function accumulator
+
+    ***************************************************************************/
+
+    public void serialize (scope SerializeDg dg) const @safe
+    {
+        serializePart(this.data[0 .. this.length], dg);
+    }
```

Comment:
I see that the 1kb only applies to data in memory, not the one on disk. It's better, but still, why not simply use a dynamic array ?

Reply:
I think it's weird, too.
I misinterpreted the meaning of `empty fields` in your last proposal.
I'll fix it.
https://github.com/bpfkorea/agora/pull/1258#discussion_r507196744